File: dossier/tui/detail.py
# Copyright © 2026-present gsfernandes81
#
# This file is part of "dossier".
#
# dossier is free software: you can redistribute it and/or modify it under the
# terms of the GNU Affero General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later version.
#
# dossier is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License along with
# dossier. If not, see <https://www.gnu.org/licenses/>.

"""Pure Rich renderer for the detail pane (the third Miller column).

Like :mod:`dossier.tui.rows`, this is widget-free and Console-testable: it takes
a :class:`~dossier.query.DocumentView` plus the resolved context the home screen
gathers (location label, supersession chain, the doc that supersedes this one)
and returns a Rich renderable the pane drops into a ``Static``.
"""

from __future__ import annotations

from collections.abc import Sequence
from datetime import date

from rich.console import Group, RenderableType
from rich.rule import Rule
from rich.text import Text

from dossier.model import Document, ExpiryStatus, FileStatus, SuggestedField, Suggestion
from dossier.query import DocumentView
from dossier.tui.glyphs import ASCII, GlyphSet

_FIELD_LABEL = {
    SuggestedField.ISSUE: "issue date",
    SuggestedField.EXPIRY: "expiry date",
    SuggestedField.NOTES: "period",
}

_EXPIRY_STYLE = {
    ExpiryStatus.EXPIRED: "bold red",
    ExpiryStatus.EXPIRING: "yellow",
    ExpiryStatus.OK: "green",
    ExpiryStatus.NONE: "dim",
}


def render_detail(
    view: DocumentView,
    *,
    location_label: str | None,
    chain: list[Document],
    superseded_by: Document | None,
    suggestions: Sequence[Suggestion] = (),
    glyphs: GlyphSet = ASCII,
) -> RenderableType:
    """Render the full detail of one document.

    ``location_label`` is the effective location already resolved to a title +
    slot; ``chain`` is the documents this one supersedes (newest replaced first);
    ``superseded_by`` is the newer document that replaced this one, if any;
    ``suggestions`` are the live field suggestions (a dim "press e to review" hint).
    """
    doc = view.document
    header = Text(doc.name or doc.id, style="bold", overflow="fold")

    facts = Text()
    if superseded_by is not None:
        _icon(facts, glyphs.link)
        facts.append("superseded by ", style="dim")
        facts.append(f"{superseded_by.name or superseded_by.id}\n", style="yellow")
    _field(facts, glyphs.location, "Location", location_label or "—")
    _field(facts, glyphs.calendar, "Issued", _fmt(doc.issue_date))
    _field(facts, glyphs.calendar, "Expires", _expiry_value(view))
    if doc.ignore_expiry:
        facts.append("expiry ignored\n", style="dim")
    _field(facts, glyphs.tag, "Tags", " ".join(doc.tags) or "—")
    _field(facts, glyphs.bundle, "Bundles", " ".join(doc.bundles) or "—")
    _field(facts, glyphs.physical, "Copies", _copies(doc))

    parts: list[RenderableType] = [header, Rule(style="dim"), facts]

    files = _files(view, glyphs)
    if files is not None:
        parts.append(files)
    if chain:
        parts.append(_chain(chain, glyphs))
    if suggestions:
        parts.append(_suggestions(suggestions, glyphs))
    if doc.notes:
        parts.append(Rule(style="dim"))
        notes = Text()
        _icon(notes, glyphs.note)
        notes.append(doc.notes)
        parts.append(notes)
    return Group(*parts)


def _icon(text: Text, icon: str) -> None:
    """Append a leading ``icon`` + space (nothing when the icon is empty/ASCII)."""
    if icon:
        text.append(f"{icon} ", style="dim")


def _field(text: Text, icon: str, label: str, value: str) -> None:
    _icon(text, icon)
    text.append(f"{label}: ", style="bold")
    text.append(f"{value}\n")


def _expiry_value(view: DocumentView) -> str:
    doc = view.document
    if doc.expiry_date is None:
        return "—"
    return f"{_fmt(doc.expiry_date)}  ({view.expiry.value})"


def _copies(doc: Document) -> str:
    marks = []
    if doc.has_physical:
        marks.append("physical")
    if doc.has_digital:
        marks.append("digital")
    return " + ".join(marks) or "—"


def _files(view: DocumentView, glyphs: GlyphSet) -> RenderableType | None:
    doc = view.document
    if not doc.files:
        return Text("No digital file linked", style="dim")
    body = Text()
    _icon(body, glyphs.digital)
    body.append("Files\n", style="bold")
    for rendition in doc.files:
        mark = glyphs.primary if rendition.primary else " "
        body.append(f" {mark} ")
        body.append(rendition.label or "file")
        body.append(f"  {rendition.path}\n", style="dim")
    if view.file is FileStatus.MISSING:
        body.append("a linked file is missing on disk\n", style="bold red")
    return body


def _suggestions(suggestions: Sequence[Suggestion], glyphs: GlyphSet) -> RenderableType:
    body = Text()
    _icon(body, glyphs.note)
    body.append(f"Suggestions ({len(suggestions)})", style="bold")
    body.append("  a accept · e review\n", style="dim")
    for suggestion in suggestions:
        label = _FIELD_LABEL[suggestion.field]
        body.append(f"  {label}: ", style="dim")
        body.append(" / ".join(suggestion.values), style="yellow")
        body.append(f"  · from {suggestion.source}\n", style="dim")
    return body


def _chain(chain: list[Document], glyphs: GlyphSet) -> RenderableType:
    body = Text()
    _icon(body, glyphs.link)
    body.append("Supersedes\n", style="bold")
    for doc in chain:
        body.append("  ← ", style="dim")
        body.append(f"{doc.name or doc.id}\n", style="dim")
    return body


def _fmt(when: date | None) -> str:
    return when.strftime("%d %b %Y") if when is not None else "—"

File: crates/ds/src/detail.rs
// Copyright © 2026-present gsfernandes81
//
// This file is part of "dossier".
//
// dossier is free software: you can redistribute it and/or modify it under the
// terms of the GNU Affero General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later version.
//
// dossier is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License along with
// dossier. If not, see <https://www.gnu.org/licenses/>.

//! The record: everything about one document on one screen.
//!
//! REWRITE-UI.md §2 and the approved phone mockup — location, expiry with its
//! standing spelled out, issue date, tags, bundles, the file list with the
//! primary marked, what it renews, and notes. The user's review call was
//! *"detail looks good for now"*, so this renders that layout and nothing more.
//!
//! **R4 made it the editing surface** (v2's Phase 4 conclusion stands: editing
//! lives in one place, not scattered across pickers). One field is editable so
//! far — the expiry, on `ctrl+e` — and the record's job during an edit is to
//! show *which* field is being typed into, since the typing itself happens on
//! the entry line at the bottom of the screen. Without that mark the two rows
//! would be talking about each other with nothing to connect them.
//!
//! The remaining verbs (`s` supersede, `b` bundle, `u` undo) arrive with the
//! slices that implement them, not before: a hint is only ever shown for
//! something that works.

use ratatui::layout::Rect;
use ratatui::text::{Line, Span};
use ratatui::widgets::Paragraph;
use ratatui::Frame;

use crate::app::Model;
use crate::layout::{truncate, wrap};
use crate::theme::{Theme, Tone};
use crate::Status;

/// Label column, so values line up under each other.
const LABEL_COLS: usize = 10;

/// Draw the record for the highlighted row.
pub fn draw(frame: &mut Frame, area: Rect, model: &Model, theme: Theme) {
    let inner = area.width.saturating_sub(2) as usize;
    let Some(doc) = model.current() else {
        frame.render_widget(
            Paragraph::new(Line::styled(" nothing selected", theme.style(Tone::Muted))),
            area,
        );
        return;
    };
    let status = model.status(doc);

    let mut lines = vec![
        Line::styled(format!(" {}", truncate(&doc.name, inner)), theme.style(Tone::Title)),
        Line::raw(""),
        field("location", &nonempty(doc.place()), inner, theme),
    ];

    // Expiry carries its standing in words next to the date: `2026-09-28` alone
    // makes the reader do the arithmetic, and the whole point of this app is
    // that nobody should have to.
    //
    // While it is being edited the label is lit, because the value on this row
    // is the *stored* one and the one being typed is three rows down on the
    // entry line. The mark is what says those two rows are about each other.
    let editing_expiry = model
        .edit
        .as_ref()
        .is_some_and(|edit| edit.doc == doc.id && edit.field == crate::edit::Field::Expiry);
    lines.push(Line::from(vec![
        if editing_expiry { marked_label("expiry", theme) } else { label("expiry", theme) },
        Span::raw(doc.expiry_date.clone().unwrap_or_else(|| "—".into())),
        Span::raw("  "),
        Span::styled(
            match status {
                Status::Expired => "! expired",
                Status::Soon => "~ expiring soon",
                Status::Ok => "  tracked",
                Status::Untracked if doc.superseded => "· superseded",
                Status::Untracked if doc.ignore_expiry => "· watch ignored",
                Status::Untracked => "· no expiry",
            }
            .to_string(),
            theme.status(status),
        ),
    ]));

    lines.push(field(
        "issued",
        &nonempty(doc.issue_date.clone().unwrap_or_default()),
        inner,
        theme,
    ));
    lines.push(field("tags", &nonempty(doc.tags.join(" ")), inner, theme));
    lines.push(field("bundles", &nonempty(doc.bundles.join(" · ")), inner, theme));

    // Files: the count, then one line each with the primary marked — the file
    // `Enter` opens is the one with the arrow, and seeing which that is matters
    // more than any other field on this screen.
    if doc.files.is_empty() {
        lines.push(field("files", "none", inner, theme));
    } else {
        let primary = doc.primary_file().map(|f| f.path.clone());
        for (i, file) in doc.files.iter().enumerate() {
            let is_primary = Some(&file.path) == primary.as_ref();
            let head =
                if i == 0 { label("files", theme) } else { Span::raw(" ".repeat(LABEL_COLS + 1)) };
            let mut spans = vec![head];
            if i == 0 {
                spans.push(Span::styled(format!("{} ", doc.files.len()), theme.style(Tone::Muted)));
            }
            spans.push(Span::styled(
                if is_primary { "▸ " } else { "  " }.to_string(),
                theme.style(Tone::Accent),
            ));
            spans.push(Span::raw(truncate(&file.path, inner.saturating_sub(LABEL_COLS + 5))));
            lines.push(Line::from(spans));
        }
    }

    if let Some(older) = &doc.supersedes {
        let title = model
            .store
            .docs
            .iter()
            .find(|d| &d.id == older)
            .map_or_else(|| older.clone(), |d| d.name.clone());
        lines.push(Line::from(vec![
            label("renews", theme),
            Span::styled(truncate(&title, inner - LABEL_COLS), theme.style(Tone::Accent)),
        ]));
    }

    if !doc.notes.is_empty() {
        lines.extend(wrapped_field("notes", &doc.notes, inner, theme));
    }

    // No widget-level wrap: everything above is already laid out to this pane's
    // width, and letting the widget wrap as well would put a continuation line
    // at the left margin, where it reads as a new field.
    frame.render_widget(Paragraph::new(lines), area);
}

/// An em dash beats a blank: an empty value and a missing field look identical
/// on screen otherwise, and only one of them is worth fixing.
fn nonempty(value: String) -> String {
    if value.is_empty() {
        "—".into()
    } else {
        value
    }
}

fn label(text: &str, theme: Theme) -> Span<'static> {
    Span::styled(format!(" {text:<LABEL_COLS$}"), theme.style(Tone::Muted))
}

/// The label of the field currently being edited.
///
/// Reverse video, which is the same texture the selected row uses and the only
/// one this app has that needs no colour and no new attribute — `SGR 4` lands
/// through the descenders on the phone's font, and `SGR 2` may be ignored
/// outright. Reverse is the one that is always exactly one cell tall.
fn marked_label(text: &str, theme: Theme) -> Span<'static> {
    Span::styled(
        format!(" {text:<LABEL_COLS$}"),
        theme.style(Tone::Accent).add_modifier(ratatui::style::Modifier::REVERSED),
    )
}

/// A one-line field: label, then the value cut to whatever the pane leaves.
fn field(name: &str, value: &str, inner: usize, theme: Theme) -> Line<'static> {
    let value_cols = inner.saturating_sub(LABEL_COLS).max(8);
    Line::from(vec![label(name, theme), Span::raw(truncate(value, value_cols))])
}

/// A field whose value is free text: wrapped, with continuations **hanging under
/// the value column** so the field still reads as one thing.
fn wrapped_field(name: &str, value: &str, inner: usize, theme: Theme) -> Vec<Line<'static>> {
    let value_cols = inner.saturating_sub(LABEL_COLS).max(8);
    let mut lines = Vec::new();
    for (i, chunk) in wrap(value, value_cols).into_iter().enumerate() {
        let head = if i == 0 { label(name, theme) } else { Span::raw(" ".repeat(LABEL_COLS + 1)) };
        lines.push(Line::from(vec![head, Span::raw(chunk)]));
    }
    lines
}
